src/ProtocolStreamWriter.php

```diff
@@ -26,22 +38,57 @@ public function __construct($output)
      */
     public function write(Message $msg)
     {
-        $data = (string)$msg;
-        $msgSize = strlen($data);
-        $totalBytesWritten = 0;
-
-        while ($totalBytesWritten < $msgSize) {
-            error_clear_last();
-            $bytesWritten = @fwrite($this->output, substr($data, $totalBytesWritten));
-            if ($bytesWritten === false) {
-                $error = error_get_last();
-                if ($error !== null) {
-                    throw new RuntimeException('Could not write message: ' . error_get_last()['message']);
-                } else {
-                    throw new RuntimeException('Could not write message');
-                }
+        // if the message queue is currently empty, register a write handler.
+        if (empty($this->messages)) {
+            Loop\addWriteStream($this->output, [$this, 'writeData']);
+        }
+
+        $promise = new Promise();
+        $this->messages[] = [
+            'message' => (string)$msg,
+            'promise' => $promise
+        ];
+        return $promise;
+    }
+
+    /**
+     * Writes pending messages to the output stream.
+     * Must be public to be able to be used as a callback.
+     *
+     * @return void
+     */
+    public function writeData()
+    {
+        $message = $this->messages[0]['message'];
+        $promise = $this->messages[0]['promise'];
+
+        error_clear_last();
+        $bytesWritten = @fwrite($this->output, $message);
+
+        if ($bytesWritten === false) {
```

Comment:
This clause would never be satisfied. If using `@`, `fwrite` will return 0 on error instead of FALSE. The whole error handling needs to be removed.